FEATURE: Fusion (EEL) core migrations for new 9.0 API (replaces `neos/rector` for Fusion)

mhsdesign opened the pull request:
Fixes the fusion part of https://github.com/neos/neos-development-collection/issues/5607 in a simple manner as a first iteration

<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

### Behaviour for 9.0

This change introduces an extended way to write flow core migrations by providing a trait which allows for exact eel replacements. For this framework code see the changes in "Neos.Fusion" and its tests.

Also this change introduces a first major user of this new migration api. The migration "20251005080230" contains all fusion  migrations concerning the content repository that currently reside in `neos/rector` and were executed with rector.

They are now part of "Neos.Neos" and can be executed as any flow core migration:

```
flow flow:core:migrate Neos.Demo --version 20251005080230
```

The comprehensive testsuite for this cr-fusion migration was copied and adjusted from `neos/rector` to ensure all replacements work as desired.


### Behaviour for Neos 8.4

Introducing a core migration for 8.4 was considered but was found to be too magic as its truly an optional step which makes the code not backwardscompatible to previous neos versions.

You can run the adjustments manually. They will not add todo comments like the 9.0 migration. It will be safe to rerun this migration and also to run the 9.0 migration via `flow:core:migrate` when updating.

```
./flow neos9preparation:preadjustfusion Neos.Demo
```


**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
WIP: Copy `EelExpressionTransformer` from `neos/rector` to "Neos.Fusion" package

and add test for `EelExpressionTransformer` to assert correct behaviour.

The afx parser modifications will be overhauled as currently the afx service breaks due to the changed AST

mhsdesign committed:
WIP: Add from and to offset for expressions in AFX ast

and introduce corresponding tests.

The positions are introduced in a way the AST is still primarily the same shape and to have little overhead during maintenance (e.g. no variables flying around).

mhsdesign committed:
FEATURE: Add first fusion migration leveraging a `FusionMigrationTrait` and thus extending flows `AbstractMigration`

Fusion migrations behave essentially to the migration framework just like any code migration.

As there is no interface for migrations - and extending the already cluttered "AbstractMigration" might be harder to follow and to test we introduce a trait which introduces new utilities like "replaceEelExpression" and more in the future.

mhsdesign committed:
FEATURE: Introduce ESCR Fusion rector rules from neos/rector

Enable comment support for migrations

Copy all inline declared rules FusionFlowQueryNodePropertyToWarningComment and FusionNodePropertyPathToWarningComment
Copy all declared fusion rules (implementing `FusionRectorInterface`) and inline their workings.

Copy all test fixtures and flatten structure

Some tests fail currently as they are no longer executed separately but with the full Fusion migration applied

mhsdesign committed:
WIP: Update fusion migration snapshots with new todo comments

mhsdesign committed:
PATCH: Simplify FusionMigrationTrait by using one `preg_replace` per file for replacements instead of multiple and re-parsing the file.

mhsdesign committed:
TASK: Add experimental replaceEelExpressionInsideFusionPath() to specify eel expression based on fusion path

mhsdesign committed:
PATCH: Fix Style CI

dlubitz commented:
I didn't had a deeper look into it yet. But I think we can't merge this into 8.4 as the core migrations are mandatory and not optional. So everyone updating to 8.4 and running core migrations would also migrate all fusion. But the backport idea was just to provide something to prepare. WDYT?

mhsdesign committed:
FEATURE: Add utility to FusionMigrationTrait to rename fusion objects

This allows to fully replace neos/rector in the fusion segment.

mhsdesign commented:
Thanks for your insights. Flow core migrations behave in mysterious ways and people use them all differently. We had the discussion where we both praised the migrations for allowing to provide them across the eco system but also we criticised how they are executed wildly without no real reason and just inject them as executed into the "composer.json".

At first was about to implement the migrations as something new - with a custom CLI - but then dialled down in favour of having something simpler and also already familiar.

The goal for the 8.4 release is to provide all new fusion APIs we migrate to and thus i find it also sensible to provide a way to automatically use them in the 8.4 branch. If there are hickups in the migration or we fear that people might be shooting them in the foot we can probably find a way to skip these migrations by default (though there is currently no way from the flow framework side).

markusguenther commented:
Thanks a bunch to both of you! When fusion is still running after the migration and is just more compatible with Neos 9.0, what would be the reason against it? The adjustment becomes a show stopper only when the rendering is completely different or you’re getting errors. But that shouldn’t be the case, right? Just my two cents, 🙈

Sebobo commented:
I think this is great to have, but we should a note to the upgrade instructions on how to skip this one if one doesn't want to migrate Fusion just yet. So adding the version string to the composer.json of the package that should get the `core:migrate` treatment.

mhsdesign commented:
cc @kitsunet @kdambekalns 

we will also include a note in the upgrade instructions that `./flow flow:core:migrate` should be reviewed and in the case for the 8.4 update we use the core migrations to identify problems with Neos 9 ahead. Some of these problems are just denoted as todo comment on the head of the file for future reference when actually updating to Neos 9. Many adjustments will be automatically applied. But also here some adjustments come with a caveat as they dont reflect the 8.3 behaviour 100% but already align to the Neos 9.0 specification.

A few examples are:
- `node.nodeType` always returns a NodeType or if removed the `Neos.Neos:FallbackNode`. In Neos 9.0 the exists no magic for the `Neos.Neos:FallbackNode` and thus the helper `Neos.Node.nodeType(node)` returns "NULL" in Neos 8.4 and 9.0.
- `node.context.currentRenderingMode` always returns the rendering mode based on the logged in user - so when viewing the page logged in in the frontend the mode is still 'inPlaces' as in the backend -> the Neos 8.4 and 9.0 `renderingMode` reports "frontend" as expected for this case
- the "live" rendering mode was renamed to "frontend" so the unlikely case of `node.context.currentRenderingMode == "live"` fails when migrated to `renderingMode.name`
- `node.context.currentSite` is rewritten to `Neos.Site.findBySiteNode(site)` which makes the assumption of "site" being present and that "currentSite" is actually the current and was not tampered with
- unlike `node.identifier` the `node.aggregateId` is now a value object in fusion, it is string-able and can be output directly but no direct strict comparison must be done `node.aggregateId == "some-id"` will not work. It has to be string casted `String.toString(node.aggregateId)`.

... some of these examples are at this point already purely academic and might never apply in real world. The Neos.Demo did survive the migration and if we find any case where it doesnt we can fix it afterwards and people can just keep using the 8.3 style.

To be fully on the safe side - which defeats the purpose of your 8.4 update as you probably want to see the weaknesses in your fusion code base and already apply most of the adjustments - you can fully skip the fusion migrations by either not running `./flow flow:core:migrate` or telling flow they have been run by inserting them into the `composer.json` which makes subsequent runs of `./flow flow:core:migrate` ignore them.

```json
{
    "extra": {
        "applied-flow-migrations": [
            "Neos.Neos-20251005080230",
            "Neos.Fusion-20251006080506"
        ]
    }
}
```

mhsdesign committed:
TASK: Restructure Fusion Migration code

dlubitz commented:
I'm still not happy with this. Let's assume someone does this migration with 8.4 but add new code afterwards which is not 9.0 compatible, you would need to re-apply this migration on upgrade to Neos 9. But you can't (except of removing it manually from compoer.json).

I see the point, that it has an advantage to "pre-migrate" your code for Neos 9, but this sould be something manual and repeatable. 

EDIT: Also adding this migration to composer.json to skip it for 8.4 would skip it also for 9.0.

mhsdesign commented:
That is certainly not ideal ... 

In that regard im not happy with the full flow migration framework and behaviour. Its sometimes a little too automatic which leads to applying migrations which are not even relevant (e.g. not a dependency) or in like in this case we have no control to skip a migration or rerun it (`--force` doesnt help but could?)

Anyway especially as i reuse next-to zero code from the migration framework but implement the file iteration and writing myself we could also just take all this code and add it to a custom cli command which could look like "./flow migratefusion:neos9compatibilty"
This would of course removes the capabilities we have for migrations like ensuring a clean git dir and committing the changes - (unless we add that back in) but it would work.
And for Neos 9.0 we could have the migrations then as this pr suggested it.

FYI rerunning the migration will also lead to duplicate `// Todo 9.0` comments in the header as we would just continuously add those over and over as the problem was not fixed. So maybe even not that wise ...

Sebobo commented:
Aren't the 8.4 Fusion migrations and 9.0 migrations two separate migrations? Meaning, even if you run this one, then add legacy code and then upgrade, you still get your Fusion migrated?

kdambekalns commented:
Running migrations again (using `--force`) would be a nice featurette to be added with this. I'd certainly prefer this over adding another CLI command…

dlubitz commented on Neos.Neos/Tests/Behavior/Features/Bootstrap/FeatureContext.php:
```diff
@@ -460,7 +460,7 @@ public function iWaitForElement($elementName)
     public function locatePath($path)
     {
-        return parent::locatePath($this->resolvePath($path));
+        // return parent::locatePath($this->resolvePath($path));
```
Is this intended?

mhsdesign replied:
yes - these are the old MINK browser tests which are not run in CI and removed in 9.0 - but that this ended up here in the diff was a hickup in github ... now its gone ;)

dlubitz reviewed (commented):
Thank you very much! +1 from my side for the PR and -1 for targeting 8.4 (see comments above). But this is just my opinion an not blocking. So feel free to merge if everyone agrees on that.

mhsdesign committed:
TASK: Introduce dedicated and optional command "neos9preparation:preadjustfusion" instead of adding it as core migration for neos 8.4

While a core migration is certainly a more beautiful way it also comes with the dark magic of being just run without control whatsoever.

Doing the partial migration in 8.4 already is totally optional - this is not imminent when using core migrate.

And when applying the core migrations to an open source package it will become incompatible with earlier releases like Neos 8.3 which is not desired.

mhsdesign committed:
PATCH: Ensure fusion migration keeps files unchanged when run again

mhsdesign commented:
Okay as discussed in todays weekly with martin, denny, seb, jon we introduce a command `flow neos9preparation:preadjustfusion Neos.Demo` instead which will not emit the todo comments. Only with Neos 9 the migration shall be turned into a real core migration with todo comments.

mhsdesign committed:
PATCH: Ignore phpstan problem with notfound class

mhsdesign committed:
PATCH: Remove `Rector` suffix in test files

mhsdesign committed:
PATCH: Remove `Rector` name in test files

mficzel commented on Neos.Neos/Tests/Unit/CodeMigrations/Version20251005080230/Fixture/FusionCacheLifetime/some_file.fusion.inc:
I remember that we did not want to add the comments in 9.4 as they cannot be addressed there. This fixture looks like they are still there

mhsdesign replied:
yes sorry for the confusion they are in there because the tests test the 9.0 upcoming flow core migration and not what is done via `neos9preparation:preadjustfusion` which will not emit comments as discussed indeed.

dlubitz reviewed (approved):
❤️
